lib/Monads/nondet/Nondet_While_Loop_Rules.thy

```diff
@@ -562,37 +583,45 @@ lemma snd_X_return[simp]:
   "snd ((A >>= (\<lambda>a. return (X a))) s) = snd (A s)"
   by (clarsimp simp: return_def bind_def split_def)
 
+lemma isr_Inr_projr:
+  "\<not> isl a \<Longrightarrow> (a = Inr b) = (b = projr a)"
+  by auto
+
 lemma whileLoopE_liftE:
   "whileLoopE C (\<lambda>r. liftE (B r)) r = liftE (whileLoop C B r)"
   apply (rule ext)
   apply (clarsimp simp: whileLoopE_def)
   apply (rule prod_eqI)
    apply (rule set_eqI, rule iffI)
     apply clarsimp
-    apply (clarsimp simp: in_bind whileLoop_def liftE_def)
-    apply (rule_tac x="b" in exI)
-    apply (rule_tac x="projr a" in exI)
+    apply (clarsimp simp: in_liftE whileLoop_def)
+    \<comment> \<open>FIXME: existential is instantiated by 'subst isr_Inr_proj' ... 'rule refl' in two lines
+    apply (rule_tac x="projr a" in exI)\<close>
+    apply (rule exI)
```

Comment:
Are we ok with leaving a schematic from `exI` like this, to be instantiated a few steps later? This was the nicest way I could find to do it that didn't require referring to a bound variable.

Reply:
Yes, I think that is fine and should be more stable than mentioning the name. The price is that it's probably a bit harder to understand what is going on, but I think for a single exI that's Ok.